### PR TITLE
Tutorials/api

### PR DESCRIPTION
**Pull request by jaimin45**
Changes According sir given !!

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Review by JainishaParmar — APPROVED** [review]

File: controllers/tutorialsControllers.js [hunk 1]
```diff
@@ -7,48 +7,52 @@ const tutorialsAll = async (req, res) => {
       createdAt: -1,
     });
     res.json(tutorials);
+    res.status(200).send();
   } catch (error) {
-    res.json({ message: error });
+    res.status(404).send();
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: We should specify error message here.

Pattern of error should be:
```
{
  message: 'not found'
}
```

Same pattern should be applied to all the error responses.

File: controllers/tutorialsControllers.js [hunk 2]
```diff
@@ -7,48 +7,52 @@ const tutorialsAll = async (req, res) => {
...
+    if (typeof published !== "undefined") update.published = published;
     const tutorialUpdate = await Tutorial.findByIdAndUpdate(
-      { _id: req.params.tutorialId },
-      { published: req.body.published },
+      { _id: req.params.Id },
       update,
-      options
+      { new: true }
     );
     res.json(tutorialUpdate);
```
Comment (kapilbarad) [inline controllers/tutorialsControllers.js]: Don't need to return anything here. Just Response with status code 203 should befine.
```
POST /tutorials - Should return 201 with newly created record.
PATCH /tutorials/:id - Should return 201 without any response.
DELETE /tutorials/:id - Should return 200 without any response.
```
CC: @ashishkava @JainishaParmar

File: routes/tutorials.js [hunk 3]
```diff
@@ -9,9 +9,9 @@ const {
 
 /* GET users listing. */
 router.get("/", tutorialsAll);
-router.get("/:tutorialId", tutorialSearchById);
+router.get("/:Id", tutorialSearchById);
 router.post("/", tutorialsCreate);
-router.put("/:tutorialId", updateTutorials);
-router.delete("/:tutorialId", deleteTutorials);
+router.patch("/:Id", updateTutorials);
```
Comment (kapilbarad) [inline routes/tutorials.js]: Parameters should be in camel case.
`Id` should be `id`, Same for the other APIs.

File: swagger.yaml [hunk 4]
```diff
@@ -49,30 +52,51 @@ paths:
...
         "200":
           description: Tutorial is updated
+        "500":
+          description: "Internal Server Error"
+    get:
+      tags:
+        - Tutorials
+      summary: Get tutorials By ID
+      responses:
+        "200":
```
Comment (kapilbarad) [inline swagger.yaml:107]: Missing 500 response for all the APIs.

File: swagger.yaml [hunk 5]
```diff
@@ -32,6 +32,9 @@ paths:
           description: OK
           schema:
             $ref: "#/definitions/Tutorial"
+        "404":
```
Comment (kapilbarad) [inline swagger.yaml]: We also need to provide definition for the error response which will be common for all the APIs.

For example:
```
"404":
  description: "Tutorial not found"
  schema:
    $ref: "#/definitions/404Response"
```

Same for the 500 status code.

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: .eslintrc.json [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    "extends": ["plugin:prettier/recommended", "airbnb-base"],
+    "parserOptions": {
+        "ecmaVersion": 2021,
+        "sourceType": "module"
+    },
+    "plugins": ["prettier", "spellcheck"],
+    "rules": {
+      "quotes": "off",
+      "new-cap": 0,
+      "no-unused-vars": "off",
```
Comment (kapilbarad) [inline .eslintrc.json]: Please remove below rules from here:
```
no-unused-vars
new-cap
comma-dangle

```

File: swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+
+  /tutorial/{id}:
+    parameters:
+      - name: id
+        in: path
+        required: true
+        description: ID of the tutorial that we want to find
```
Comment (kapilbarad) [inline swagger.yaml]: No need of the `that we want to find` as we've this parameter common for DELETE and GET apis

File: swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+
+    patch:
+      summary: Update tutorial with give ID
+      tags:
+        - Tutorials
+      responses:
+        "201":
```
Comment (kapilbarad) [inline swagger.yaml]: Patch Should have 204 response without response body Please apply the same in controller.

File: swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+          schema:
+            $ref: "#/definitions/404Response"
+        "500":
+          description: "Internal Server Error"
+          schema:
+            $ref: "#/definitions/500Response"
+    get:
+      tags:
+        - Tutorials
+      summary: Get tutorials By ID
```
Comment (kapilbarad) [inline swagger.yaml]: Typo: `Get tutorial by ID`

File: swagger.yaml [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    post:
+      summary: Create tutorials in system
+      tags:
+        - Tutorials
+      parameters:
+        - name: Tutorials
+          in: body
+          description: User that we want to create
+          schema:
+            $ref: "#/definitions/Tutorial"
```
Comment (kapilbarad) [inline swagger.yaml]: While creating new Tutorial, body only needs to title and description nothing else. so we can't reference `#/definitions/Tutorial` here. We need to create another definition for example `TutorialCreateBody` with only title and description fields.

CC: @JainishaParmar @ashishkava

File: swagger.yaml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        - Tutorials
+      responses:
+        "201":
+          description: Tutorial is updated
+          schema:
+            $ref: "#/definitions/Tutorial"
+        "400":
+          description: "Invalid Tutorial Id"
+          schema:
+            $ref: "#/definitions/404Response"
```
Comment (kapilbarad) [inline swagger.yaml]: Wrong ref.

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by kapilbarad — APPROVED** [review]

**Review by kapilbarad — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by JainishaParmar — APPROVED** [review]

**Comment by jaimin45** [comment]
Features/APIs

**Pull request merged**